code/code/disc/disc_alchemy.cc

```diff
@@ -2208,28 +2209,109 @@ int castGalvanize(TBeing *caster, TObj *obj)
   return FALSE;
 }
 
-int ethrealGate(TBeing *caster, TObj *obj, int, short bKnown)
+struct gateRoomT {
+      int roomnum;
+      const char *name;
+};
+
+gateRoomT gateRooms[] =
 {
-  return SPELL_FALSE;
+  {15346, "grimhaven"},
+  {11000, "cloud"},
+  {5700, "auria"},
+  {6953, "circle"},
+  {12883, "volcano"},
+};
+
+const int NUM_GATE_ROOMS = 5;
+
+
+int etherealGate(TBeing *caster, const char * portalroom, int, short bKnown)
+{
+  char buf[256];
+  TPerson *tPerson = dynamic_cast<TPerson *>(caster);
+  int i, location=0;
+  int num_charges = 20;
+  sstring sbuf;
+
+
+  for(i=0;i<NUM_GATE_ROOMS;++i){
+    if(is_abbrev(portalroom, gateRooms[i].name)){
+      location=gateRooms[i].roomnum;
+    }
+  }
+
+  TRoom * rp = real_roomp(location);
+
+  if (!rp || !location) {
+    caster->sendTo("You can't seem to open a gate to that location.\n\r");
+    return SPELL_FAIL;
+  }
+
+  if (caster->roomp->isRoomFlag(ROOM_NO_PORTAL) && !caster->isImmortal()) {
+    caster->sendTo("You can't seem to escape the defenses of this area.\n\r");
+    act("Nothing seems to happen.", FALSE, caster, NULL, NULL, TO_ROOM);
+    return SPELL_FAIL;
+  }
+
+  if (caster->bSuccess(bKnown,caster->getPerc(),SPELL_ETHER_GATE)) {
+    TPortal * tmp_obj = new TPortal(rp);
+    tmp_obj->setPortalNumCharges(num_charges);
+    tmp_obj->obj_flags.decay_time = 5;
+    tmp_obj->name = "portal gate";
+    tmp_obj->shortDescr = "<c>A <B>sw<W>ir<1><b>li<C>ng <1><c>ethereal gate<1>";
```

Comment:
I thought about that but I already run into issues because the SQL files here aren't in sync with the mud. I think the code is better if it doesn't rely on loading specific vnums.

This code does look kind of ugly to me too though. I feel like it would be better to pass these as parameters to the constructor rather than create the object and then modify a bunch of stuff. 

In go I'd probably create a struct to pass to the constructor but does that sound like proper the proper way in C++?